cdmtaskservice/s3/client.py

```diff
@@ -0,0 +1,207 @@
+        except (ValueError, EndpointConnectionError) as e:
+            raise S3ClientConnectError(f"s3 connect failed: {e}") from e
+        except ResponseParserError as e:
+            # TODO TEST logging
+            # TODO LOGGING figure out how logging is going to work
+            logging.getLogger(__name__).error(
+                f"Unable to parse response from S3:\n{e}\n")
+            raise S3ClientConnectError(
+                f"s3 response from the server at {self._url} was not parseable. "
+                + "See logs for details"
+            ) from e
+        except ClientError as e:
+            code = e.response["Error"]["Code"]
+            if code == "SignatureDoesNotMatch":
+                raise S3ClientConnectError("s3 access credentials are invalid")
+            if code == "404":
+                raise S3PathError(f"The path '{path}' was not found on the s3 system") from e
+            if code == "AccessDenied" or code == "403":  # why both? Both 403s
+                if not path:
+                    raise S3ClientConnectError(
+                        "Access denied to list buckets on the s3 system") from e
+                # may need to add other cases here
+                raise S3PathError(f"Access denied to path '{path}' on the s3 system") from e
+            # no way to test this since we're trying to cover all possible errors in tests
+            logging.getLogger(__name__).error(
+                f"Unexpected response from S3. Response data:\n{e.response}\nTraceback:\n{e}\n")
+            raise S3UnexpectedError(str(e)) from e
+
+
+    async def get_object_meta(self, path: str) -> S3ObjectMeta:
+        """
+        Get metadata about an object.
+        
+        path - the path of the object in s3, starting with the bucket.
+        """
+        buk, key = _validate_and_split_path(path)
+        async def head(client):
+            return await client.head_object(Bucket=buk, Key=key, PartNumber=1)
+        res = await self._run_command(head, path=path.strip())
+        size = res["ContentLength"]
+        part_size = None
+        if "PartsCount" in res:
+            part_size = size
+            content_range = res["ResponseMetadata"]["HTTPHeaders"]["content-range"]
+            size = int(content_range.split("/")[1])
+        return S3ObjectMeta(
+            path=path,
+            e_tag = res["ETag"].strip('"'),
+            size=size,
+            part_size=part_size
+        )
+
+
+def _validate_and_split_path(path: str) -> (str, str):
+    if not path or not path.strip():
+        raise S3PathError("An s3 path cannot be null or a whitespace string")
+    parts = [s.strip() for s in path.split("/", 1) if s.strip()]
+    if len(parts) != 2:
+        raise S3PathError(
+            f"path '{path.strip()}' must start with the s3 bucket and include a key")
+    _validate_bucket_name(parts[0])
+    return parts[0], parts[1]
+
+
+def _validate_bucket_name(bucket_name: str):
```

Comment:
Do we need to validate bucket name ourselves? Will S3client return reasonable error message if we don't?

Reply:
I want to fail fast when possible rather than making a call, and I'm pretty sure I remember putting this in because handling the error was annoying.

Reply:
👍 The logic does look solid. Hopefully there won't be an edge case that we might overlook. But I guess then the client will raise an error eventually.

Reply:
Right, worst case is we throw a harder to understand error.